File: caves-of-zircon.java/src/main/java/rltut/ai/PlayerAi.java
package rltut.ai;

import org.hexworks.zircon.api.data.impl.Position3D;
import rltut.algorithm.FieldOfView;
import rltut.entities.Item;
import rltut.world.Tile;

import java.util.List;

public class PlayerAi extends CreatureAi {

	private List<String> messages;
	private FieldOfView fov;
	
	public PlayerAi(Creature creature, List<String> messages, FieldOfView fov) {
		super(creature);
		this.messages = messages;
		this.fov = fov;
	}

	public void onEnter(Position3D pos, Tile tile){
		if (tile.isGround()){
			creature.position = pos;
			
			Item item = creature.item(creature.position);
			if (item != null)
				creature.notify("There's a " + creature.nameOf(item) + " here.");
			
		} else if (tile.isDiggable()) {
			creature.dig(pos);
		}
	}
	
	public void onNotify(String message){
		messages.add(message);
	}
	
	public boolean canSee(Position3D pos) {
		return fov.isVisible(pos);
	}
	
	public void onGainLevel(){
	}

	public Tile rememberedTile(Position3D pos) {
		return fov.tile(pos);
	}
}

File: caves-of-zircon.kotlin/src/main/kotlin/org/hexworks/cavesofzircon/ai/PlayerAI.kt
package org.hexworks.cavesofzircon.ai

import org.hexworks.cavesofzircon.blocks.Creature
import org.hexworks.cavesofzircon.blocks.GameBlock
import org.hexworks.cavesofzircon.world.World
import org.hexworks.cobalt.datatypes.extensions.map
import org.hexworks.zircon.api.data.Position

class PlayerAI(private val player: Creature,
               private val world: World) : CreatureAI {

    override fun onEnter(position: Position, block: GameBlock): Boolean {
        return when {
            block.isGround() -> true
            block.isDiggable() -> {
                player.dig(position)
                false
            }
            block.hasCreature() -> {
                world.creature(position).map {
                    player.attack(it)
                }
                false
            }
            else -> {
                throw UnsupportedOperationException("Can't comprehend block: $block")
            }
        }
    }
}
